### ISSUE TITLE
dead store to static variable not eliminated

### ISSUE THREAD

**Issue by fefe17** [llvm/llvm-project#55699]
Consider this test code:

```
int dummy1()
{
  static int removeme = 0;
  if (removeme)
  {
      return 0;
  }
  removeme = 1;
  return 0;
}

int dummy2()
{
  static int removeme = 0;
  if (!removeme)
    removeme = 1;
  return 0;
}

int dummy3()
{
  static int removeme = 0;
  removeme = 1;
  return 0;
}
```

My expectation was: emit "return 0" for all three.
Observed behavior: emit conditional + store and then return 0 for dummy1 and dummy2, emit only return 0 for dummy3.

Am I missing something? Why isn't this dead store eliminated?

**Comment by Meinersbur** [comment]
dummy3 is optimized because there are only stores to `removeme` but no reads.

dummy1 and dummy2 need to take into account that those functions can be called concurrently. For instance if it was
```
int dummy1_concurrent()
{
  static int removeme = 0;
  if (removeme)
  {
      return 1; // instead of return 0
  }
  removeme = 1;
  return 0;
}
```
two concurrently called functions could return 0 until `removeme` is assigned. Also involves the target platform's memory model. A singleton pattern would require some synchronization mechanism, like the compiler emits if `removeme` had an initializer. 

To optimize, the compiler would need to deduce that the effect of the value of `removeme` only has an effect on `removeme` itself. This is probably rare in non-synthetic programs.

**Comment by fefe17** [comment]
Fair point but we weren't talking about a code snippet where concurrent execution could change the outcome.

In fact, most people I talked to about this were convinced potential races in concurrent environments were to blame.
I was under the impression that the C++ 11 memory model at least makes it clear that data races are undefined behavior and the compiler can thus assume they don't happen when generating code. If you want code that works in the face of concurrent data races, you need to use atomics or a mutex.

**Comment by dwblaikie** [comment]
I /think/ the optimizations would be valid, but haven't been implemented/probably aren't a priority given the complexity involved (the somewhat circular reasoning that the write is dead because the read only causes the write, etc
